### PR TITLE
Stack property track

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
This draft is intended to keep track of what is currently supported by the naive approach.

1. Size of the function is determined by `framet`
2. Stack size is determined by `--stack-limit`

Current limitations:

1. Unions sizes are not being calculated correctly
2. Inner scopes are not being handled.

### CODE REVIEW

File: src/goto-symex/builtin_functions.cpp [hunk 1]
```diff
@@ -626,6 +626,37 @@ void goto_symext::intrinsic_get_thread_state(
   symex_assign(code_assign2tc(call.ret, flag_expr), true);
 }
 
+void goto_symext::intrinsic_get_stack_size(
+  const code_function_call2t &call,
+  reachability_treet &art)
+{
+  // Get all variables
+  unsigned size = 0;
+  ns.get_context().foreach_operand([this,&size](const symbolt &s) {
```
Comment (mikhailramalho) [inline src/goto-symex/builtin_functions.cpp]: Here, you are getting all the variables in the program, including global variables and variables defined in other functions.

If you want to track the stack size while it is growing, simply:
1. Add a new attribute to framet, which tracks frame information.
2. Change symex_decl to update this field whenever a new variable is declared. You can also add a new option to trigger an property violation if a limit is reached.
3. You can also decrease the stack size when scoped variables go out of scope, but I'm not sure this is the right behaviour, e.g.:
```
void foo()
{
  int a, b;
  {
    int a, b;
  }
  // What is the stack size here? Four integers or two?
}
```
I'm assuming it's two or we could blow the stack by declaring variable inside for-loops, but it's better to check the standard.

File: src/goto-symex/symex_stack.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#include <util/expr_util.h>
+#include <util/irep2.h>
+
+void goto_symext::process_stack_size(expr2tc &expr)
+{
+  code_decl2t &decl_code = to_code_decl2t(expr);
+  cur_state->top().frame_info.add_decl(decl_code);
+  if(stack_limit != -1)
+  {
+    static unsignedbv_type2tc u32type(32);
```
Comment (lucasccordeiro) [inline src/goto-symex/symex_stack.cpp]: Shall we set it to 32 or 64 according to the option given in the command-line?

File: src/goto-symex/symex_stack.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#include <cassert>
+#include <goto-symex/goto_symex.h>
+#include <util/expr_util.h>
+#include <util/irep2.h>
+
+void goto_symext::process_stack_size(expr2tc &expr)
+{
+  code_decl2t &decl_code = to_code_decl2t(expr);
+  cur_state->top().frame_info.add_decl(decl_code);
+  if(stack_limit != -1)
```
Comment (lucasccordeiro) [inline src/goto-symex/symex_stack.cpp]: Why "-1"?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

File: src/goto-symex/symex_stack.cpp (hunk 3 above)
Comment (rafaelsamenezes) [reply src/goto-symex/symex_stack.cpp]: The value of `stack_limit`comes from an `atol` and since the default value for the parameter is -1 this checks if the user has changed it.

A better approach could be done when using BigInt, but we have plans of replacing it: https://github.com/esbmc/esbmc/issues/181

File: src/goto-symex/goto_symex_state.h [hunk 4]
```diff
@@ -185,6 +185,18 @@ class goto_symex_statet
...
+      std::vector<std::pair<std::string, unsigned>> variables;
+      unsigned total = 0;
+
+      void add_decl(code_decl2tc s)
+      {
+        total += s->type->get_width();
+        variables.emplace_back(s->value.as_string(), s->type->get_width());
+      }
+    } frame_info;
+
```
Comment (mikhailramalho) [inline src/goto-symex/goto_symex_state.h:205]: You can use statet::framet instead of creating a new struct:
https://github.com/esbmc/esbmc/blob/master/src/goto-symex/goto_symex_state.h#L148

Also, do you really need the keep track of the variable name?

File: src/goto-symex/goto_symex.h [hunk 5]
```diff
@@ -711,6 +711,12 @@ class goto_symext
    */
   void replace_nondet(expr2tc &expr);
 
+  /**
+   * Process a block adding the width of each symbol into the stack length
+   * @param expr Expr to search for symbols.
+   */
+  void process_stack_size(expr2tc &expr);
+
```
Comment (mikhailramalho) [inline src/goto-symex/goto_symex.h]: I think this should be part of statet::framet.

File: src/goto-symex/symex_other.cpp [hunk 6]
```diff
@@ -62,8 +62,9 @@ void goto_symext::symex_decl(const expr2tc code)
   replace_dynamic_allocation(code2);
   replace_nondet(code2);
   dereference(code2, dereferencet::READ);
+  process_stack_size(code2);
 
-  const code_decl2t &decl_code = to_code_decl2t(code2);
+  code_decl2t &decl_code = to_code_decl2t(code2);
```
Comment (mikhailramalho) [inline src/goto-symex/symex_other.cpp:80]: Why did you remove the const?

File: src/goto-symex/symex_stack.cpp (hunk 2 above)
Comment (mikhailramalho) [reply src/goto-symex/symex_stack.cpp]: Why static? The `irep2tc`s are shared_pointers.

Also, you should use the `type_pool` methods to get default types. I suggest using `get_uint64` regardless of the arch.

File: src/goto-symex/symex_stack.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  {
+    static unsignedbv_type2tc u32type(32);
+    BigInt f_size(cur_state->top().frame_info.total);
+    static BigInt s_size(stack_limit);
```
Comment (mikhailramalho) [inline src/goto-symex/symex_stack.cpp]: No need for it to be static, `stack_limit` is already an attribute of the class.

File: src/goto-symex/symex_stack.cpp [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    constant_int2tc function_irep2(u32type, f_size);
+    constant_int2tc limit_irep2(u32type, s_size);
+
+    lessthanequal2tc check(function_irep2, limit_irep2);
+    claim(check, "Stack limit property was violated");
+    ;
```
Comment (mikhailramalho) [inline src/goto-symex/symex_stack.cpp]: superfluous `;`.

Comment (lucasccordeiro) [reply src/goto-symex/symex_stack.cpp]: fixed.

File: src/goto-symex/symex_stack.cpp (hunk 7 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_stack.cpp]: Fixed.

File: src/goto-symex/symex_other.cpp (hunk 6 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_other.cpp:80]: Fixed.

File: src/goto-symex/builtin_functions.cpp (hunk 1 above)
Comment (rafaelsamenezes) [reply src/goto-symex/builtin_functions.cpp]: For 3 we could use the llvm bitcode as a base. All `alloca` calls are at the top of the function.

The major issue is how to map some optimizations that involves registers e.g. `mem2reg`

File: src/goto-symex/goto_symex_state.h (hunk 4 above)
Comment (rafaelsamenezes) [reply src/goto-symex/goto_symex_state.h:205]: This is a inner struct inside framet.

About the variable name... not now. But I am guessing that in the future it will be useful for counterexample purposes.

File: src/goto-symex/builtin_functions.cpp (hunk 1 above)
Comment (lucasccordeiro) [reply src/goto-symex/builtin_functions.cpp]: I have tried to look for information about the stack size when scoped variables go out of scope, but I was unable to find anything relevant to this discussion in https://web.cs.dal.ca/~vlado/pl/C_Standard_2011-n1570.pdf and http://www2.open-std.org/JTC1/SC22/WG14/www/abq/c17_updated_proposed_fdis.pdf. I might have overlooked these documents.

File: src/goto-symex/goto_symex.h (hunk 5 above)
Comment (rafaelsamenezes) [reply src/goto-symex/goto_symex.h]: I put in `goto_symext`because the claim will need to be stack size of the current execution. Currently, each frame only contains the size of its function, the correct claim will contain all frames so far.

File: src/goto-symex/goto_symex_state.h (hunk 4 above)
Comment (mikhailramalho) [reply src/goto-symex/goto_symex_state.h:205]: But you still don't need the vector of variables or the struct:
1. `total` can be an attribute of framet (an it needs a better name).
2. Since you always call `add_decl` when creating the claim, you can simply add the name to the claim:
```
void goto_symext::process_stack_size(expr2tc &expr)
{
  const code_decl2t &decl_code = to_code_decl2t(expr);
  cur_state->top().frame_info.add_decl(decl_code); /* I. REMOVE THIS */
  if(stack_limit != 0)
  {
    unsignedbv_type2tc u32type(32);
    BigInt f_size(cur_state->top().frame_info.total);
    BigInt s_size(stack_limit);
    constant_int2tc function_irep2(u32type, f_size);
    constant_int2tc limit_irep2(u32type, s_size);

    lessthanequal2tc check(function_irep2, limit_irep2);

    /* II. ADD THE NAME OF THE VARIABLE TO THE CLAIM */
    claim(check, "Stack limit property was violated when declarin " + to_decl_code(expr).value.as_string());
  }
}
```

I've added the two comments in uppercase so they are easier to spot.

File: src/goto-symex/goto_symex.h (hunk 5 above)
Comment (mikhailramalho) [reply src/goto-symex/goto_symex.h]: I don't get it: the new option `stack-limit` defines a global stack limit, i.e., the combined stacks should be smaller than this number?

If that's the case, you can move this to statet::framet and add a new argument: 
```
void process_stack_size(expr2tc &expr, goto_symex_statet *cur_state);
```
and in there you can get the stack limit and call `cur_state->claim`.

Comment (lucasccordeiro) [reply src/goto-symex/goto_symex.h]: I have moved `process_stack_size` to `goto_symex_statet::framet`, as suggested by @mikhailramalho. However, I have used a different strategy to deal with the `claim` and `stack_limit`. In particular, I added to `symex_decl` the following implementation:

```
  if(stack_limit != 0)
  {
    claim(
      (cur_state->top().process_stack_size(code2, stack_limit)),
      "Stack limit was violated");
  }
```

File: src/goto-symex/goto_symex_state.h (hunk 4 above)
Comment (lucasccordeiro) [reply src/goto-symex/goto_symex_state.h:205]: The second comment is really nice. 

Now we get the variable name in the claim:

```
Counterexample:

State 1 file main.c line 8 function main thread 0
----------------------------------------------------
Violated property:
  file main.c line 8 function main
  Stack limit property was violated when declaring xPtr

VERIFICATION FAILED
```
For this program:
```
  1 int main() {
  2   int y = 5;  // 4 * 8 = 32
  3   int *yPtr;  // 32
  4   // Expected total here: 32+32= 64.
  5 
  6   yPtr = &y;
  7 
  8   int *xPtr; // 32
  9   // The total now is 64 + 32 = 96.
 10   return 0;
 11 }
```

Now, in `symex_decl`, we have the follwoing:

```
  if(stack_limit != 0)
  {
    const std::string pretty_name =
      to_code_decl2t(code).value.as_string().substr(
        to_code_decl2t(code).value.as_string().find_last_of('@') + 1);

    claim(
      (cur_state->top().process_stack_size(code2, stack_limit)),
      "Stack limit property was violated when declaring " + pretty_name);
  }
```

Comment (lucasccordeiro) [reply src/goto-symex/goto_symex_state.h:205]: I agree with @mikhailramalho 

I have removed the struct and variables vector from framet.

I have also changed the variable `total` to `stack_frame_total`.

**Comment by lucasccordeiro** [comment]
@mikhailramalho: This PR is ready to be reviewed again.

File: src/goto-symex/builtin_functions.cpp (hunk 1 above)
Comment (lucasccordeiro) [reply src/goto-symex/builtin_functions.cpp]: In this particular example given by @mikhailramalho

```
void foo()
{
  int a, b;
  {
    int a, b;
  }
  // What is the stack size here? Four integers or two?
}
```

LLVM produces:

$clang -emit-llvm -c stack.c
$llvm-dis stack.bc

```
; Function Attrs: noinline nounwind optnone ssp uwtable
define void @main() #0 {
  %1 = alloca i32, align 4
  %2 = alloca i32, align 4
  %3 = alloca i32, align 4
  %4 = alloca i32, align 4
  ret void
}
```

ESBMC is following the same behaviour of LLVM:

$esbmc stack.c --32 --stack-limit 127

```
Counterexample:

State 1 file stack.c line 5 function main thread 0
----------------------------------------------------
Violated property:
  file stack.c line 5 function main
  Stack limit property was violated when declaring b

VERIFICATION FAILED
```

$esbmc stack.c --32 --stack-limit 128

```
...
Generated 4 VCC(s), 0 remaining after simplification (0 assignments)
BMC program time: 0.000s

VERIFICATION SUCCESSFUL
```

File: src/goto-symex/symex_other.cpp [hunk 9]
```diff
@@ -62,8 +62,19 @@ void goto_symext::symex_decl(const expr2tc code)
   replace_dynamic_allocation(code2);
   replace_nondet(code2);
   dereference(code2, dereferencet::READ);
+  stack_limit = atol(options.get_option("stack-limit").c_str());
```
Comment (mikhailramalho) [inline src/goto-symex/symex_other.cpp]: No need to get the `stack_limit` here again, there is a member in the class with the same information.

File: src/goto-symex/symex_other.cpp [hunk 10]
```diff
@@ -62,8 +62,19 @@ void goto_symext::symex_decl(const expr2tc code)
...
+      to_code_decl2t(code).value.as_string().substr(
+        to_code_decl2t(code).value.as_string().find_last_of('@') + 1);
+
+    claim(
+      (cur_state->top().process_stack_size(code2, stack_limit)),
+      "Stack limit property was violated when declaring " + pretty_name);
+  }
 
-  const code_decl2t &decl_code = to_code_decl2t(code2);
+  code_decl2t &decl_code = to_code_decl2t(code2);
```
Comment (mikhailramalho) [inline src/goto-symex/symex_other.cpp]: Why did you remove the const here?

File: src/goto-symex/symex_other.cpp [hunk 11]
```diff
@@ -62,8 +62,19 @@ void goto_symext::symex_decl(const expr2tc code)
   replace_dynamic_allocation(code2);
   replace_nondet(code2);
   dereference(code2, dereferencet::READ);
+  stack_limit = atol(options.get_option("stack-limit").c_str());
+  if(stack_limit > 0)
+  {
+    const std::string pretty_name =
```
Comment (mikhailramalho) [inline src/goto-symex/symex_other.cpp]: This looks dreadful (not your fault).

We need a separate function in another PR to clean up the names when writing these claims (memory leak claims and dead code claims also suffer from the same problem).

File: src/goto-symex/symex_other.cpp [hunk 12]
```diff
@@ -62,8 +62,19 @@ void goto_symext::symex_decl(const expr2tc code)
   replace_dynamic_allocation(code2);
   replace_nondet(code2);
   dereference(code2, dereferencet::READ);
+  stack_limit = atol(options.get_option("stack-limit").c_str());
+  if(stack_limit > 0)
```
Comment (mikhailramalho) [inline src/goto-symex/symex_other.cpp:68]: Some comments here would be helpful.

File: src/goto-symex/symex_other.cpp (hunk 10 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_other.cpp]: Fixed.

File: src/goto-symex/symex_other.cpp (hunk 9 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_other.cpp]: Fixed.

File: src/goto-symex/symex_other.cpp (hunk 11 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_other.cpp]: I agree with @mikhailramalho 

Could we add the pretty name to our irep2 node?

File: src/goto-symex/symex_other.cpp (hunk 12 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_other.cpp:68]: Implemented as suggested.

**Comment by mikhailramalho** [comment]
It seems mostly ok now, but what about handling symex_dead? 

It should change the stack calculation when the variables go out of scope, right?

**Comment by lucasccordeiro** [comment]
This PR is ready to be reviewed.

File: src/goto-symex/symex_function.cpp [hunk 13]
```diff
@@ -589,7 +589,7 @@ bool goto_symext::make_return_assignment(expr2tc &assign, const expr2tc &code)
   return false;
 }
 
-void goto_symext::symex_return()
+void goto_symext::symex_return(expr2tc code)
```
Comment (mikhailramalho) [inline src/goto-symex/symex_function.cpp]: This should be a const reference.

File: src/util/pretty.h [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <util/irep2.h>
+#include <util/irep2_expr.h>
+
+inline std::string get_pretty_name(const expr2tc code)
+{
+  return to_code_decl2t(code).value.as_string().substr(
+    to_code_decl2t(code).value.as_string().find_last_of('@') + 1);
+}
+
```
Comment (mikhailramalho) [inline src/util/pretty.h:22]: good!

**Review by mikhailramalho — APPROVED** [review]

File: src/goto-symex/symex_function.cpp (hunk 13 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_function.cpp]: Fixed.
